### ISSUE TITLE
Enabling the InferSendableFromCaptures upcoming feature flag can break code that involves C functions

### ISSUE THREAD

**Issue by groue** [swiftlang/swift#73313]
### Description

When I enable the upcoming feature flag `InferSendableFromCaptures` (from [SE-0418](https://github.com/apple/swift-evolution/blob/main/proposals/0418-inferring-sendable-for-methods.md)), the Swift 6 compiler emits errors:

> A C function pointer can only be formed from a reference to a 'func' or a literal closure.

Those errors are not emitted when `InferSendableFromCaptures` is not enabled.

### Reproduction

```swift
import SQLite3

func f(_ cFunction: @convention(c) (OpaquePointer?) -> UnsafeMutablePointer<CChar>?) { }

func g() {
  // A C function pointer can only be formed from a reference to a 'func' or a literal closure
  f(sqlite3_expanded_sql)
}
```

### Expected behavior

No compiler error.

### Environment

```
$ /Library/Developer/Toolchains/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-04-22-a.xctoolchain/usr/bin/swift --version
Apple Swift version 6.0-dev (LLVM 94447cf359772fc, Swift f36098a14a0d48b)
Target: arm64-apple-macosx14.0
```

### Additional information

I've been using this technique for many years, with great success, inspired by [this post on the Swift forums](https://forums.swift.org/t/c-interoperability-combinations-of-library-and-os-versions/14029/4).

**Comment by jckarter** [comment]
https://github.com/apple/swift/pull/73436 should fix this.

**Comment by groue** [comment]
Hello, with swift-6.0-DEVELOPMENT-SNAPSHOT-2024-05-26-a the error with C function is indeed solved. Now instead I get compiler crashes. I'll open other issues.

**Comment by groue** [comment]
The bug is back in Xcode 16.0 beta (16A5171c).

**Comment by xedin** [comment]
I don’t think the beta has the fix but 6.0 branch does, which is what we use to verify issues.

**Comment by groue** [comment]
Fixed in Xcode 16.0 beta 3 (16A5202i) 👍
